### PR TITLE
docs: add CONTRIBUTING.md with RFC and testing guidelines

### PR DESCRIPTION
**Pull request by Matovidlo**
## Description

**Linear**: N/A (docs-only chore)

### Change Type

- [ ] Major (breaking changes, significant new features)
- [ ] Minor (new features, enhancements, backward compatible)
- [x] Patch (bug fixes, small improvements, no new features)

### Summary

Adds a `CONTRIBUTING.md` that codifies the RFC requirement and testing expectations
that have been implicit in the project so far.

Key rules:

- **RFC gate**: An RFC in `feature_spec/<name>/RFC.md` is required — and must be reviewed
  before implementation — whenever the change introduces a **new MCP tool**, a **new
  architectural concept**, **new end-to-end behavior**, or **fully new functionality** with
  no existing hook to extend. The trigger is the *nature* of the change, not the line count.
  Bug fixes, refactoring, and enhancements to existing tools are exempt.
- **Bug fix testing**: Every fix must include a regression test. An E2E scenario (integration
  test or manual demo) in the PR description is strongly recommended.
- **Feature testing**: Unit tests required with minimal mocking; E2E / integration tests
  required for externally visible behavior; prefer extending existing parametrized tests
  over adding new test functions.
- **All PRs**: version bump, `uv.lock` sync, `tox` green, self-review.

Existing RFC documents are referenced as examples.

## Testing

No code changes — documentation only.

## Checklist

- [x] Self-review completed
- [ ] Unit tests added/updated (if applicable)
- [ ] Integration tests added/updated (if applicable)
- [x] Project version bumped according to the change type (if applicable) — N/A for docs-only
- [x] Documentation updated (if applicable)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+Some changes need design agreement before implementation starts. The signal is the **nature** of the change, not the line count — a new function parameter with many test cases is fine without an RFC; a new tool with two lines of glue code is not.
+
+### When an RFC is required
+
+| Change type | RFC required? |
+|---|---|
+| New MCP tool | **Yes** |
+| New end-to-end behavior or data flow | **Yes** |
+| New architectural concept (new client, new session model, new transport layer) | **Yes** |
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md:22]: **Fixed** ✅ in commit 98211f6a5d06d5c6f87b3e44ca93843147c286e7

Changed `|---|---|` to `| --- | --- |` — the spaced form is unambiguous across all GFM renderers.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
+# Contributing to Keboola MCP Server
+
+## Before You Start
+
+Check the [Linear board](https://linear.app/keboola) for an existing issue, or create one.
+Branch names must start with the Linear issue ID (e.g. `AI-1234-short-description`).
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md:6]: **Fixed** ✅ in commit 98211f6a5d06d5c6f87b3e44ca93843147c286e7

Added the release branch exception: "Exception: release branches use `release/vX.Y.Z` and do not require a Linear issue ID." — mirrors CLAUDE.md exactly.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+See [CLAUDE.md](CLAUDE.md) for the full git workflow, versioning rules, and testing setup.
+
+---
+
+## RFC Requirement
+
+Some changes need design agreement before implementation starts. The signal is the **nature** of the change, not the line count — a new function parameter with many test cases is fine without an RFC; a new tool with two lines of glue code is not.
+
+### When an RFC is required
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md:16]: **Fixed** ✅ in commit 98211f6a5d06d5c6f87b3e44ca93843147c286e7

Renamed the checklist section from "Large features (≥ 200 lines)" to "Features requiring an RFC". The line-count heuristic is gone; the RFC table above is the single source of truth for when an RFC is needed.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+### Bug fixes
+
+- [ ] Regression test added that reproduces the bug
+- [ ] E2E scenario described in PR description (wrong behavior → correct behavior)
+
+### Large features (≥ 200 lines)
+
+- [ ] RFC in `feature_spec/<feature-name>/RFC.md` reviewed and agreed on **before** this PR
+- [ ] RFC linked in the PR description
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md]: **Fixed** ✅ in commit 98211f6a5d06d5c6f87b3e44ca93843147c286e7

Updated the PR description to describe the RFC gate in terms of change nature (new tool, new architecture, new end-to-end behavior) — no line-count threshold anywhere now.

**Review by padak — COMMENTED** [review]
Thanks for putting this together — the document itself is solid and the first round of feedback is well addressed. Before approval I'd flag three small things to reconcile, since this PR is meant to codify the rules and benefits from being internally consistent.

### 1. `isort` missing from the listed `tox` checks

`pyproject.toml` defines:

```
env_list = ["python", "black", "isort", "flake8", "check-tools-docs"]
```

So the default `tox` run actually includes `isort`. CONTRIBUTING.md currently lists only `pytest + black + flake8 + check-tools-docs` in two places:

- Line 96 (general testing guidelines)
- Line 114 (PR checklist item)

This matches the two still-open Copilot comments. (Side note: the existing `CLAUDE.md` has the same omission — out of scope here, but worth a follow-up.)

### 2. PR template ↔ CONTRIBUTING.md mismatch on the version bump

- `.github/pull_request_template.md`: `Project version bumped according to the change type (if applicable)`
- CONTRIBUTING.md (this PR), line 113: `pyproject.toml version bumped (patch / minor / major per CLAUDE.md)` — no `(if applicable)`

`CLAUDE.md § Versioning` says *"Every PR must bump `pyproject.toml` version"* with docs explicitly in the patch tier, so the unconditional wording in CONTRIBUTING.md is the correct one. The fix is probably to drop `(if applicable)` from the PR template rather than to weaken CONTRIBUTING.md.

### 3. This PR doesn't follow its own rule

`git diff main..HEAD -- pyproject.toml` is empty — the version stays at `1.60.0`. The PR description marks the version bump as *"N/A for docs-only"*, but per `CLAUDE.md` (and per the rule this PR is codifying), docs PRs should land a patch bump.

Suggested fix: bump `pyproject.toml` from `1.60.0` to `1.60.1`, run `uv lock`, and commit both.

---

None of these are big — once they're resolved, happy to approve.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  scenarios. Use `@pytest.mark.parametrize` with a new parameter axis.
+- Only add test cases that cover genuinely new behavior — do not duplicate cases already
+  handled by existing parametrize entries.
+
+### General testing guidelines
+
+These apply to all change types:
+
+- Run `tox` before pushing. CI runs the same checks (pytest, black, flake8,
+  check-tools-docs) and will fail if any of them fail.
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md]: **Fixed** ✅ in commit d4da7d46d8528f265ebaefde418395690b39e6a7

Added `isort` to the parenthetical on line 95 — `pyproject.toml` `env_list` includes `isort` and CI runs it via `uv run tox`, so the list was incomplete. Fixed in the commit above.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+Use the [PR template](.github/pull_request_template.md). Make sure every item below is
+addressed before requesting review.
+
+### All PRs
+
+- [ ] Branch name starts with the Linear issue ID
+- [ ] Commit messages start with the Linear issue ID
+- [ ] `pyproject.toml` version bumped (patch / minor / major per CLAUDE.md)
+- [ ] `uv.lock` synced (`uv lock`)
+- [ ] `tox` passes locally (pytest + black + flake8 + check-tools-docs)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md]: **Fixed** ✅ in commit d4da7d46d8528f265ebaefde418395690b39e6a7

Added `isort` to the PR checklist tox line — same root cause as the comment above: `env_list` includes `isort` so omitting it from the checklist would leave contributors surprised by CI failures.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+Use the [PR template](.github/pull_request_template.md). Make sure every item below is
+addressed before requesting review.
+
+### All PRs
+
+- [ ] Branch name starts with the Linear issue ID
+- [ ] Commit messages start with the Linear issue ID
+- [ ] `pyproject.toml` version bumped (patch / minor / major per CLAUDE.md)
+- [ ] `uv.lock` synced (`uv lock`)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply CONTRIBUTING.md:113]: **Fixed** ✅ in commit d4da7d46d8528f265ebaefde418395690b39e6a7

Removed `(if applicable)` from the version bump item in `.github/pull_request_template.md`. `CLAUDE.md` says "Every PR must bump" unconditionally, and `CONTRIBUTING.md` already reflected that — the PR template was the outlier. Both now read as unconditional.

**Comment by padak** [comment]
## LGTM

Docs-only change is clean: version bump (1.60.0 → 1.60.1) and `uv.lock` sync are in order, the referenced RFC paths (`feature_spec/branched_storage_support/RFC.md`, `feature_spec/project_info_branch_context/RFC.md`) exist in the repo, and the tox checklist in CONTRIBUTING.md is actually more accurate than what is in CLAUDE.md today (correctly includes `isort`). Nothing blocking — a few non-blocking suggestions below.

### Non-blocking suggestions

1. **Broken anchor links to CLAUDE.md** (`CONTRIBUTING.md:5` and `:97`)
   ```markdown
   See [CLAUDE.md](CLAUDE.md) for the full git workflow, ...
   See [CLAUDE.md § Testing](CLAUDE.md) for venv setup ...
   ```
   Both jump to the top of the file — `§ Testing` in the label is purely cosmetic. If the anchor is meant to work, use `(CLAUDE.md#testing)` (and a matching slug for the git-workflow section). GitHub generates slugs as lowercase + dashes.

2. **Release-branch exception in the commit-message rule**
   Line 6 mentions the exception for branch names only. CLAUDE.md says release branches do *not* carry a Linear issue ID, so the author of a release PR cannot honestly tick „Commit messages start with the Linear issue ID" in the checklist. Worth adding the analogous exception to the commit-message rule.

3. **„Version bumped" checklist item without `(if applicable)`**
   The change from `... (if applicable)` to no qualifier is stricter and aligns with CLAUDE.md („Every PR must include a `pyproject.toml` version bump"). This PR satisfies it (patch bump even though docs-only). Just worth confirming the intent is *truly always* — including PRs that touch only `.github/` or `docs/`.

4. **Duplicated tox step list**
   The checklist line `tox passes locally (pytest + black + isort + flake8 + check-tools-docs)` repeats the list already given in „General testing guidelines". Not wrong, but if CI gains e.g. `mypy`, it has to be updated in two places. Suggestion: keep the enumeration in the section, leave only `tox passes locally` in the checklist.

5. **„Enhancement vs. fully new functionality" edge in the RFC table**
   The rows „Enhancement to existing tool / function (new param, extended response)" vs. „Fully new functionality with no existing hook to extend" are hard to disambiguate without context. A short concrete example below the table (e.g. „adding a `branch_id` parameter to `find_component_id` = enhancement; adding a `list_oauth_credentials` tool = new functionality") would save review-time debates.

6. **Follow-up suggestion**: sync CLAUDE.md § Testing so it also mentions `isort` (it currently lists only four envs), so the two docs do not contradict each other.

**Review by padak — APPROVED** [review]
LGTM — see inline comment for non-blocking suggestions.
